hedera-node/hedera-networkadmin-service-impl/src/main/java/com/hedera/node/app/service/networkadmin/impl/handlers/FreezeHandler.java

```diff
@@ -54,77 +61,136 @@ public FreezeHandler() {
-
-                // FREEZE_ONLY requires a valid start_time
-            case FREEZE_ONLY -> verifyFreezeStartTimeIsInFuture(freezeTxn, txnId.transactionValidStart());
-
-                // PREPARE_UPGRADE requires valid update_file and file_hash values
-            case PREPARE_UPGRADE -> verifyUpdateFileAndHash(freezeTxn, specialFileStore);
-
-                // FREEZE_UPGRADE and TELEMETRY_UPGRADE require a valid start_time and valid update_file and
-                // file_hash values
-            case FREEZE_UPGRADE, TELEMETRY_UPGRADE -> {
-                verifyFreezeStartTimeIsInFuture(freezeTxn, txnId.transactionValidStart());
-
-                // from proto specs, it looks like update file not required for FREEZE_UPGRADE and TELEMETRY_UPGRADE
-                // but specs aren't very clear
-                // current code in FreezeTransitionLogic checks for the file in specialFiles
-                // so we will do the same
-                verifyUpdateFileAndHash(freezeTxn, specialFileStore);
-            }
-
-                // FREEZE_ABORT does not require any additional checks
-            case FREEZE_ABORT -> {
-                // do nothing
-            }
+        if (freezeType == UNKNOWN_FREEZE_TYPE) {
+            throw new PreCheckException(ResponseCodeEnum.INVALID_FREEZE_TRANSACTION_BODY);
         }
-
-        // no need to add any keys to the context because this transaction does not require any signatures
-        // it must be submitted by an account with superuser privileges, that is checked during ingest
     }
 
     @Override
     public void handle(@NonNull final HandleContext context) throws HandleException {
-        throw new UnsupportedOperationException("Not implemented yet");
+        requireNonNull(context);
+        final var txn = context.body();
+        final NetworkAdminServiceConfig adminServiceConfig =
+                context.configuration().getConfigData(NetworkAdminServiceConfig.class);
+        final ReadableSpecialFileStore specialFileStore = context.readableStore(ReadableSpecialFileStore.class);
+        final WritableFreezeStore freezeStore = context.writableStore(WritableFreezeStore.class);
+
+        final FreezeTransactionBody freezeTxn = txn.freezeOrThrow();
+        final FileID updateFileNum =
+                freezeTxn.updateFile(); // only some freeze types require this, it may be null for others
+
+        validateSemantics(freezeTxn, specialFileStore, updateFileNum);
+
+        final FreezeUpgradeActions upgradeActions = new FreezeUpgradeActions(adminServiceConfig, freezeStore);
+        final Timestamp freezeStartTime = freezeTxn.startTime(); // may be null for some freeze types
+        final Instant freezeStartTimeInstant = freezeStartTime == null
+                ? null
+                : Instant.ofEpochSecond(freezeStartTime.seconds(), freezeStartTime.nanos());
+
+        // @todo('Issue #6761') - the below switch returns a CompletableFuture, need to feed this to ExecutorService
```

Comment:
We don't want to send the future to an executor service (you cannot actually do that).  We want to pass an executor service to these methods, and provide that executor service as a parameter to the `runAsync()` call that returns the CompletableFuture.
That task (passing in the ExecutorService) can be done immediately after this PR.

Reply:
I updated the comment